### PR TITLE
Multiparty

### PR DESCRIPTION
**Pull request by eric-feng-2011**
This PR adds code to generate gRPC clients and servers for Opaque.
This introduces separating the client from the Opaque driver.

### CODE REVIEW

File: src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala [hunk 1]
```diff
@@ -249,6 +254,11 @@ object Utils extends Logging {
     }
   }
 
+  def initServer() : Unit = {
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala:257]: Do we use this?

File: src/main/scala/edu/berkeley/cs/rise/opaque/benchmark/Benchmark.scala [hunk 2]
```diff
@@ -46,12 +49,17 @@ object Benchmark {
     //   if (spark.sparkContext.isLocal) 1 else spark.sparkContext.defaultParallelism
 
     // Warmup
-    LogisticRegression.train(spark, Encrypted, 1000, 1)
-    LogisticRegression.train(spark, Encrypted, 1000, 1)
+//    LogisticRegression.train(spark, Encrypted, 1000, 1)
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/benchmark/Benchmark.scala:52]: Why did you comment this out?

**Review by chester-leung — CHANGES_REQUESTED** [review]
Looks great! Thank as well for merging the latest version of master into this branch. Can you split the merging of master and your gRPC work into two PRs? Perhaps we can create another PR into the `multiparty` branch that just merges the latest version of master. Once we merge that in this PR should update to only show your changes.

Couple high level comments:
* please remove all commented out print statements you added for debugging
* could you please remove the auto-generated gRPC files and add some documentation on how to generate them? ideally we may also generate them in build.sbt
* can you please provide a more detailed description of what the changes enable in the PR description?

File: src/main/scala/edu/berkeley/cs/rise/opaque/benchmark/Benchmark.scala (hunk 2 above)
Comment (eric-feng-2011) [reply src/main/scala/edu/berkeley/cs/rise/opaque/benchmark/Benchmark.scala:52]: Commented it out previously while figuring if I can just run everything in benchmark.scala and never reverted. Reverted now.

File: src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala (hunk 1 above)
Comment (eric-feng-2011) [reply src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala:248]: No. Removed.

File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueClient.scala [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  def getRA(name: String): Unit = {
+    val request = Ra.RARequest
+	.newBuilder()
+	.setName(name)
+	.build()
+    try {
+      val response = blockingStub.getRA(request)
+      if (response.getSuccess()) {
+	// Verify the RA report is correct and return the client key if so
+        val msg2 = serviceProvider.ProcessEnclaveReport(response.getReport().toByteArray())
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueClient.scala:83]: Can you make the variable names here more descriptive? e.g. is msg2 the attestation report from the enclave?

File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueClient.scala [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      }
+    }
+    catch {
+      case e: StatusRuntimeException =>
+        logger.log(Level.WARNING, "RPC failed: {0}", e.getStatus)
+    }
+  }
+
+  def sendTestQuery(): Unit = {
+    val testQuery = "SELECT * from Test"
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueClient.scala:104]: Have you tested other queries as well?

File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      System.err.println("*** server shut down")
+      
+      // Drop test table
+      spark.catalog.dropTempView("Test")
+    }
+  }
+
+  private def stop(): Unit = {
+    if (server != null) {
+      server.shutdown()
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala:66]: Does this shutdown function call the code on lines 54-60 above?

File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+import org.apache.spark.sql.SparkSession
+
+import implicits._
+
+/**
+ * [[https://github.com/grpc/grpc-java/blob/v0.15.0/examples/src/main/java/io/grpc/examples/helloworld/HelloWorldServer.java]]
+ */
+object OpaqueServer {
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala:20]: If my understanding is correct, this is the gRPC server listening at the driver, and the gRPC server essentially relays gRPC requests from the client and forwards to the driver. If so, can we rename this to something like `OpaqueRPCListener"

File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  private class GreeterImpl extends GreeterGrpc.GreeterImplBase {
+    override def sayHello(req: Ra.HelloRequest, responseObserver: io.grpc.stub.StreamObserver[Ra.HelloReply]) = {
+      val reply = Ra.HelloReply.newBuilder().setMessage("Hello " + req.getName()).build();
+      responseObserver.onNext(reply);
+      responseObserver.onCompleted();
+    }
+
+    override def getRA(req: Ra.RARequest, responseObserver: io.grpc.stub.StreamObserver[Ra.RAReply]) = {
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala:83]: Presuming that this function is just an RPC wrapper around the `GenerateReport` function that enables the client to remotely call `GenerateReport`, can we rename this to something like `relayGenerateReport`?

File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      val reply = Ra.RAReply.newBuilder()
+	.setReport(byteString)
+	.setSuccess(true)
+	.build();
+
+      responseObserver.onNext(reply);
+      responseObserver.onCompleted();
+    }
+
+    override def sendQuery(req: Ra.QueryRequest, responseObserver: io.grpc.stub.StreamObserver[Ra.QueryReply]) = {
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala:99]: Rename to `relayQuery`

File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        reply = Ra.QueryReply.newBuilder()
+	  .setSuccess(false)
+	  .build()
+      } finally {
+        responseObserver.onNext(reply);
+        responseObserver.onCompleted();
+      }
+    }
+
+    override def sendKey(req: Ra.KeyRequest, responseObserver: io.grpc.stub.StreamObserver[Ra.KeyReply]) = {
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala:122]: Later on once we add certificates etc I think this function will do more than just relay the key. Can we rename to `relayFinishAttestation` or something like that?

File: src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala [hunk 10]
```diff
@@ -277,7 +278,8 @@ object Utils extends Logging {
     val cipherText = data.drop(GCM_IV_LENGTH)
     val cipher = Cipher.getInstance("AES/GCM/NoPadding", "SunJCE")
     cipher.init(Cipher.DECRYPT_MODE, cipherKey, new GCMParameterSpec(GCM_TAG_LENGTH * 8, iv))
-    cipher.doFinal(cipherText)
+    val ret = cipher.doFinal(cipherText)
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala:281]: Was this changed for any reason?

File: src/main/scala/edu/berkeley/cs/rise/opaque/benchmark/Benchmark.scala [hunk 11]
```diff
@@ -46,7 +46,7 @@ object Benchmark {
     //   if (spark.sparkContext.isLocal) 1 else spark.sparkContext.defaultParallelism
 
     // Warmup
-    LogisticRegression.train(spark, Encrypted, 1000, 1)
+    LogisticRegression.train(spark, Insecure, 1000, 1)
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/benchmark/Benchmark.scala:49]: Revert

File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+      responseObserver.onNext(reply);
+      responseObserver.onCompleted();
+    }
+
+    override def sendQuery(req: Ra.QueryRequest, responseObserver: io.grpc.stub.StreamObserver[Ra.QueryReply]) = {
+      var reply = Ra.QueryReply.newBuilder().build();
+      try {
+	// Prepare items for setting data reply
+        val result = spark.sql(req.getSqlQuery())
```
Comment (chester-leung) [inline src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala:103]: Have you also tried testing this for loading saved tables, in particular in EncryptedSource format? i.e. to simulate the context, you can create an encrypted dataframe in Opaque and save it to disk (all at the server), then try to load that dataframe and manipulate it over gRPC.

**Review by chester-leung — CHANGES_REQUESTED** [review]
Thanks for the changes! Some more comments.

**Comment by chester-leung** [comment]
Can you also get the build to autogenerate the gRPC files?

**Pull request closed without merging**